File: src/test/java/leets/leenk/domain/feed/domain/service/LinkedUserSaveServiceTest.java
package leets.leenk.domain.feed.domain.service;

import leets.leenk.domain.feed.domain.entity.Feed;
import leets.leenk.domain.feed.domain.entity.LinkedUser;
import leets.leenk.domain.feed.domain.repository.LinkedUserRepository;
import leets.leenk.domain.feed.test.FeedTestFixture;
import leets.leenk.domain.feed.test.LinkedUserTestFixture;
import leets.leenk.domain.feed.test.UserTestFixture;
import leets.leenk.domain.user.domain.entity.User;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;

import static org.mockito.BDDMockito.then;

@ExtendWith(MockitoExtension.class)
public class LinkedUserSaveServiceTest {
    @Mock
    private LinkedUserRepository linkedUserRepository;

    @InjectMocks
    private LinkedUserSaveService linkedUserSaveService;

    @Test
    @DisplayName("linkedUserSave 테스트")
    void linkedUserSave() {
        // given
        User user = UserTestFixture.createUser(1L, "me");
        Feed feed = FeedTestFixture.createFeed(1L, user);
        LinkedUser linkedUser = LinkedUserTestFixture.createLinkedUser(null, user, feed);

        // when
        linkedUserSaveService.save(linkedUser);

        // then
        then(linkedUserRepository).should().save(linkedUser);
    }

    @Test
    @DisplayName("saveAll 테스트")
    void saveAll() {
        // given
        User user = UserTestFixture.createUser(1L, "me");
        Feed feed = FeedTestFixture.createFeed(1L, user);
        LinkedUser linkedUser = LinkedUserTestFixture.createLinkedUser(null, user, feed);
        List<LinkedUser> linkedUsers = List.of(linkedUser);

        // when
        linkedUserSaveService.saveAll(linkedUsers);

        // then
        then(linkedUserRepository).should().saveAll(linkedUsers);
    }
}

File: src/test/kotlin/leets/leenk/domain/feed/domain/service/LinkedUserSaveServiceTest.kt
package leets.leenk.domain.feed.domain.service

import io.kotest.core.spec.style.BehaviorSpec
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import leets.leenk.domain.feed.domain.repository.LinkedUserRepository
import leets.leenk.domain.feed.test.FeedTestFixture
import leets.leenk.domain.feed.test.LinkedUserTestFixture
import leets.leenk.domain.feed.test.UserTestFixture

class LinkedUserSaveServiceTest :
    BehaviorSpec({
        val linkedUserRepository = mockk<LinkedUserRepository>(relaxed = true)
        val linkedUserSaveService = LinkedUserSaveService(linkedUserRepository)

        Given("연결된 사용자가 주어졌을 때") {
            val user = UserTestFixture.createUser(1L, "me")
            val feed = FeedTestFixture.createFeed(1L, user)
            val linkedUser = LinkedUserTestFixture.createLinkedUser(1L, user, feed)

            every { linkedUserRepository.save(any()) } returns linkedUser

            When("연결된 사용자를 저장하면") {
                linkedUserSaveService.save(linkedUser)

                Then("linkedUserRepository의 save 메서드가 호출되어야 한다") {
                    verify { linkedUserRepository.save(linkedUser) }
                }
            }
        }

        Given("연결된 사용자 목록이 주어졌을 때") {
            val user = UserTestFixture.createUser(1L, "me")
            val feed = FeedTestFixture.createFeed(1L, user)
            val linkedUser = LinkedUserTestFixture.createLinkedUser(1L, user, feed)
            val linkedUsers = listOf(linkedUser)

            When("연결된 사용자 목록을 저장하면") {
                linkedUserSaveService.saveAll(linkedUsers)

                Then("linkedUserRepository의 saveAll 메서드가 호출되어야 한다") {
                    verify { linkedUserRepository.saveAll(linkedUsers) }
                }
            }
        }
    })
